core/src/main/java/org/opensearch/sql/planner/physical/HighlightOperator.java

```diff
@@ -0,0 +1,121 @@
+import lombok.Getter;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.opensearch.sql.common.utils.StringUtils;
+import org.opensearch.sql.data.model.ExprTupleValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.expression.DSL;
+import org.opensearch.sql.expression.Expression;
+import org.opensearch.sql.expression.ReferenceExpression;
+import org.opensearch.sql.expression.env.Environment;
+
+/**
+ * HighlightOperator evaluates the  {@link HighlightOperator#highlight} to put result
+ * into the output. Highlight fields in input are matched to the appropriate output.
+ * Direct mapping between input and output, as well as partial mapping is made
+ * dependent on highlight expression.
+ *
+ */
+@EqualsAndHashCode
+public class HighlightOperator extends PhysicalPlan {
+  @Getter
+  private final PhysicalPlan input;
+  @Getter
+  private final Expression highlight;
+
+  public HighlightOperator(PhysicalPlan input, Expression highlight) {
+    this.input = input;
+    this.highlight = highlight;
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitHighlight(this, context);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return input.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    ExprValue inputValue = input.next();
+    Pair<String, ExprValue> evalMap = mapHighlight(inputValue.bindingTuples());
+
+    if (STRUCT == inputValue.type()) {
+      ImmutableMap.Builder<String, ExprValue> resultBuilder = new ImmutableMap.Builder<>();
+      Map<String, ExprValue> tupleValue = ExprValueUtils.getTupleValue(inputValue);
+      for (Map.Entry<String, ExprValue> valueEntry : tupleValue.entrySet()) {
+        resultBuilder.put(valueEntry);
+      }
+      resultBuilder.put(evalMap);
+      return ExprTupleValue.fromExprValueMap(resultBuilder.build());
+    } else {
+      return inputValue;
+    }
+  }
+
+  /**
+   * Evaluate the expression in the {@link HighlightOperator#highlight} with {@link Environment}.
+   * @param env {@link Environment}
+   * @return The mapping of reference and {@link ExprValue} for expression.
+   */
+  private Pair<String, ExprValue> mapHighlight(Environment<Expression, ExprValue> env) {
+    String osHighlightKey = "_highlight";
+    if (!highlight.toString().contains("*")) {
+      osHighlightKey += "." + StringUtils.unquoteText(highlight.toString());
+    }
+
+    ReferenceExpression osOutputVar = DSL.ref(osHighlightKey, STRING);
+    ExprValue value = osOutputVar.valueOf(env);
+
+    // In the event of multiple returned highlights and wildcard being
+    // used in conjunction with other highlight calls, we need to ensure
+    // only wildcard regex matching is mapped to wildcard call.
+    if (StringUtils.unquoteText(highlight.toString()).matches("(.+\\*)|(\\*.+)")
```

Comment:
Is this regex only matches strings begining or ending with `*`?
I.e., will it match `B*o*d*y`?

Reply:
It will not match `B*o*d*y`, yes it must start or end with a star. As well in your example opensearch would not return any value so this if statement would fail the `and` section of this if statement.

Reply:
I'd like to double-check that. From looking at the code, OpenSearch does support `*` in the middle of the expression.

Highlight eventually uses [HighlightPhase](https://github.com/opensearch-project/OpenSearch/blob/4ed09955e7013405f21a959a044f131d3bf224f1/server/src/main/java/org/opensearch/search/fetch/subphase/highlight/HighlightPhase.java#L133), which uses OpenSearch's Regex class which [supports](https://github.com/opensearch-project/OpenSearch/blob/4ed09955e7013405f21a959a044f131d3bf224f1/server/src/main/java/org/opensearch/common/regex/Regex.java#L110) something like `B*o*d*y`. It was not the best example, but I think `B*dy` will match.